public/app/features/explore/Logs/LogsTable.tsx

```diff
@@ -74,93 +80,199 @@ export const LogsTable: React.FunctionComponent<Props> = (props) => {
 
-      // remove fields that should not be displayed
+      // remove hidden fields
+      transformations.push(...removeHiddenFields(dataFrame));
+      let labelFilters = buildLabelFilters(columnsWithMeta, logsFrame);
+
+      // Add the label filters to the transformations
+      const transform = getLabelFiltersTransform(labelFilters);
+      if (transform) {
+        transformations.push(transform);
+      }
 
-      const hiddenFields = separateVisibleFields(dataFrame, { keepBody: true, keepTimestamp: true }).hidden;
-      hiddenFields.forEach((field: Field, index: number) => {
-        transformations.push({
-          id: 'organize',
-          options: {
-            excludeByName: {
-              [field.name]: true,
-            },
-          },
-        });
-      });
       if (transformations.length > 0) {
-        const [transformedDataFrame] = await lastValueFrom(transformDataFrame(transformations, [dataFrame]));
-        setTableFrame(prepareTableFrame(transformedDataFrame));
+        const transformedDataFrame = await lastValueFrom(transformDataFrame(transformations, [dataFrame]));
+        const tableFrame = prepareTableFrame(transformedDataFrame[0]);
+        setTableFrame(tableFrame);
       } else {
         setTableFrame(prepareTableFrame(dataFrame));
       }
     };
     prepare();
-  }, [prepareTableFrame, logsFrames, logsSortOrder]);
+  }, [columnsWithMeta, logFrameRaw, logsSortOrder, prepareTableFrame]);
 
   if (!tableFrame) {
     return null;
   }
 
+  const onCellFilterAdded = (filter: AdHocFilterItem) => {
+    if (!props.onClickFilterLabel || !props.onClickFilterOutLabel) {
+      return;
+    }
+    const { value, key, operator } = filter;
+    if (operator === FILTER_FOR_OPERATOR) {
+      props.onClickFilterLabel(key, value);
+    }
+
+    if (operator === FILTER_OUT_OPERATOR) {
+      props.onClickFilterOutLabel(key, value);
+    }
+  };
+
   return (
     <Table
       data={tableFrame}
       width={width}
-      height={getTableHeight(props.logsFrames)}
+      onCellFilterAdded={onCellFilterAdded}
+      height={props.height}
       footerOptions={{ show: true, reducer: ['count'], countRows: true }}
     />
   );
+}
+
+const isFieldFilterable = (field: Field, logsFrame?: LogsFrame | undefined) => {
+  if (!logsFrame) {
+    return false;
+  }
+  if (logsFrame.bodyField.name === field.name) {
+    return false;
+  }
+  if (logsFrame.timeField.name === field.name) {
+    return false;
+  }
+  // @todo not currently excluding derived fields from filtering
+
+  return true;
 };
+
+// TODO: explore if `logsFrame.ts` can help us with getting the right fields
+// TODO Why is typeInfo not defined on the Field interface?
+function extractFieldsAndExclude(dataFrame: DataFrame) {
+  return dataFrame.fields
+    .filter((field: Field & { typeInfo?: { frame: string } }) => {
+      return (
+        (field.typeInfo?.frame === 'json.RawMessage' && field.name === 'labels') ||
```

Comment:
I haven't seen the `typeInfo` used before and it is not documented/the types don't expect it. So I am wondering how reliable it is and if we can't just use `field.type === FieldType.other` as bellow?

Reply:
Great question, I don't know